### PR TITLE
Allow searching through inventory by fields

### PR DESCRIPTION
**Pull request by nkotaa**
# Description
As mentioned in #109 , this PR adds the following:
1. Searching inventory through fields. For instance
   ``` console
   $ search -n milk -c >5 -l 2 -e 29-12-2024
   ```
   prints the first 2 items with name "milk", cost price >$5.00 expiring before 29th December 2024.
2. Improved search command parser
3. [For developers] Separate `SearchAssistant` class for calls to search the inventory

# To do
- [x] Add more tests to SearchCommandParser
- [ ] Update documentation to reflect new command syntax
- [ ] [Improvements] Case-insensitive search and price range search as suggested by @nur-haziq 
- [ ] [Enhancement] Highlight fields matched in printed results

### CODE REVIEW

File: src/main/java/seedu/binbash/inventory/SearchAssistant.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+            .limit(numberOfResults)
+            .collect(Collectors.toList());
+        return this;
+    }
+
+    public SearchAssistant searchBySalePrice(double salePriceField) {
+        foundItems = (ArrayList<Item>) foundItems.stream()
+            .filter(item -> (salePriceField < 0) ?
+                    item instanceof RetailItem && ((RetailItem) item).getItemSalePrice() < (-1 * salePriceField) :
+                    salePriceField == 0 ||
```
Comment (nur-haziq) [inline src/main/java/seedu/binbash/inventory/SearchAssistant.java:57]: Am I correct to say that when `salePriceField` == 0, the original list would be returned, which is the intended behaviour for when salePrice isn't specified in the search command?

**Review by nur-haziq — COMMENTED** [review]
LGTM. Code looks clean and readable. 👍 

There are many methods created, so do let us know if you need help creating the relevant JUnit tests for them.

File: src/main/java/seedu/binbash/inventory/SearchAssistant.java (hunk 1 above)
Comment (nkotaa) [reply src/main/java/seedu/binbash/inventory/SearchAssistant.java:64]: That is correct. This behavior is made more explicit in my latest commit.

File: src/main/java/seedu/binbash/inventory/SearchAssistant.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        this.numberOfResults = numberOfResults;
+    }
+
+    public ArrayList<Item> getFoundItems() {
+        return foundItems;
+    }
+
+    public SearchAssistant searchByName(String nameField) {
+        foundItems = (ArrayList<Item>) foundItems.stream()
+            .filter(item -> item.getItemName().contains(nameField))
```
[the remark this answers is not in the document]
Comment (nkotaa) [reply src/main/java/seedu/binbash/inventory/SearchAssistant.java:31]: Thanks for the idea.

File: src/main/java/seedu/binbash/inventory/SearchAssistant.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        foundItems = foundItems.stream()
+            .filter(item -> item.getItemDescription().contains(descriptionField))
+            .limit(numberOfResults)
+            .collect(Collectors.toCollection(ArrayList::new));
+        return this;
+    }
+
+    public SearchAssistant searchByCostPrice(double costPriceField) {
+        foundItems = (ArrayList<Item>) foundItems.stream()
+            .filter(item -> (costPriceField < 0) ? item.getItemCostPrice() < (-1 * costPriceField) :
```
[the remark this answers is not in the document]
Comment (nkotaa) [reply src/main/java/seedu/binbash/inventory/SearchAssistant.java:51]: That indeed better reflects real world usage. Right now arguments passed in to price options are of the form <{price}. Do you think they should be updated to <={price} as well? I.e
```console
$ search -s <=0.5 -n pen
```
I am open to other markers as well.

File: src/main/java/seedu/binbash/command/SearchCommand.java [hunk 4]
```diff
@@ -1,24 +1,63 @@
...
+    }
+
+    public void setExpiryDateField(LocalDate expiryDateField) {
+        this.expiryDateField = expiryDateField;
+    }
+
+    public void setNumberOfResults(int numberOfResults) {
+        this.numberOfResults = numberOfResults;
+    }
+
```
Comment (XavierLiau34) [inline src/main/java/seedu/binbash/command/SearchCommand.java:47]: You used proper and consistent formatting in your code. Well done!

File: src/main/java/seedu/binbash/parser/AddCommandParser.java [hunk 5]
```diff
@@ -39,29 +40,15 @@ public AddCommand parse(String[] commandArgs) throws ParseException {
...
         String itemName = String.join(" ", commandLine.getOptionValues("name"));// Allow multiple arguments
         String itemDescription = String.join(" ", commandLine.getOptionValues("description"));
-        double itemCostPrice;
-        try {
-            itemCostPrice = Double.parseDouble(commandLine.getOptionValue("cost"));
-        } catch (NumberFormatException e) {
-            throw new ParseException("Item cost price must be a number!");
-        }
+        double itemCostPrice = TypeHandler.createNumber(
+                commandLine.getOptionValue("cost-price")).doubleValue();
```
Comment (XavierLiau34) [inline src/main/java/seedu/binbash/parser/AddCommandParser.java:44]: Nice Simplification of Methods

File: src/test/java/seedu/binbash/inventory/SearchAssistantTest.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        String foundItemName = foundItems.get(0).getItemName();
+        Assertions.assertTrue(foundItemName.contains("banana"));
+    }
+
+    @Test
+    public void searchByDescription_sarasa_success() {
+        ArrayList<Item> foundItems = searchAssistant.searchByDescription("sarasa")
+            .getFoundItems(1);
+        String foundItemDescription = foundItems.get(0).getItemDescription();
+        Assertions.assertTrue(foundItemDescription.contains("sarasa"));
```
Comment (XavierLiau34) [inline src/test/java/seedu/binbash/inventory/SearchAssistantTest.java:49]: Assertions are used to make the code defensive. Nice

**Review by XavierLiau34 — APPROVED** [review]
Overall, LGTM. Code is functional.

**Pull request merged**